torch/csrc/dynamo/eval_frame.c

```diff
@@ -0,0 +1,578 @@
+
+  PyFrameObject* shadow = PyFrame_New(tstate, code, frame->f_globals, NULL);
+  if (shadow == NULL) {
+    return NULL;
+  }
+
+  PyObject** fastlocals_old = frame->f_localsplus;
+  PyObject** fastlocals_new = shadow->f_localsplus;
+
+  for (Py_ssize_t i = 0; i < nlocals_old; i++) {
+    Py_XINCREF(fastlocals_old[i]);
+    fastlocals_new[i] = fastlocals_old[i];
+  }
+
+  for (Py_ssize_t i = 0; i < ncells + nfrees; i++) {
+    Py_XINCREF(fastlocals_old[nlocals_old + i]);
+    fastlocals_new[nlocals_new + i] = fastlocals_old[nlocals_old + i];
+  }
+
+  PyObject* result = eval_frame_default(tstate, shadow, throw_flag);
+  Py_DECREF(shadow);
+  return result;
+}
+
+static PyObject* _custom_eval_frame_shim(
+    PyThreadState* tstate,
+    PyFrameObject* frame,
+    int throw_flag) {
+  // Shims logic into one of three states. Can probably be refactored into a
+  // single func, later:
+  //  - None: disables TorchDynamo
+  //  - False: run-only mode (reuse existing compiles)
+  //  - Python callable(): enables TorchDynamo
+  PyObject* callback = eval_frame_callback_get();
+
+  if (callback == Py_None) {
+    return eval_frame_default(tstate, frame, throw_flag);
+  }
+
+  return _custom_eval_frame(tstate, frame, throw_flag, callback);
+}
+
+static PyObject* _custom_eval_frame(
+    PyThreadState* tstate,
+    PyFrameObject* frame,
+    int throw_flag,
+    PyObject* callback) {
+  DEBUG_TRACE(
+      "begin %s %s %i %i %i %i",
+      name(frame),
+      PyUnicode_AsUTF8(frame->f_code->co_filename),
+      frame->f_lineno,
+      frame->f_lasti,
+      frame->f_iblock,
+      frame->f_executing);
+  CacheEntry* extra = get_extra(frame->f_code);
+  if (extra == SKIP_CODE || (callback == Py_False && extra == NULL)) {
+    DEBUG_TRACE("skip %s", name(frame));
+    return eval_frame_default(tstate, frame, throw_flag);
+  }
+
+  // TODO(jansel): investigate directly using the "fast" representation
+  if (PyFrame_FastToLocalsWithError(frame) < 0) {
+    DEBUG_TRACE("error %s", name(frame));
+    return NULL;
+  }
+
+  // A callback of Py_False indicates "run only" mode, the cache is checked, but
+  // we never compile.
+  if (callback == Py_False) {
+    DEBUG_TRACE("In run only mode %s", name(frame));
+    PyCodeObject* cached_code = lookup(extra, frame->f_locals);
+    if (cached_code != NULL) {
+      // used cached version
+      DEBUG_TRACE("cache hit %s", name(frame));
+      return eval_custom_code(tstate, frame, cached_code, throw_flag);
+    } else {
+      DEBUG_TRACE("cache miss %s", name(frame));
+      return eval_frame_default(tstate, frame, throw_flag);
+    }
+  }
+  DEBUG_CHECK(PyDict_CheckExact(frame->f_locals));
+  DEBUG_CHECK(PyDict_CheckExact(frame->f_globals));
+  DEBUG_CHECK(PyDict_CheckExact(frame->f_builtins));
+
+  // We don't run the current custom_eval_frame behavior for guards.
+  // So we temporarily set the callback to Py_None to drive the correct behavior
+  // in the shim.
+  eval_frame_callback_set(Py_None);
```

Comment:
Missing decref of the current one and incref of None (in the error case we never replace put it back)
We can also yolo it (None is immortal) but we should fix callback refcount on error still.